File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/level/ClientLevel.java
package pize.tests.voxelgame.client.level;

import pize.tests.voxelgame.Minecraft;
import pize.tests.voxelgame.main.chunk.storage.HeightmapType;
import pize.tests.voxelgame.main.level.Level;
import pize.tests.voxelgame.client.block.Blocks;
import pize.tests.voxelgame.client.chunk.ClientChunk;

import static pize.tests.voxelgame.main.chunk.ChunkUtils.getChunkPos;
import static pize.tests.voxelgame.main.chunk.ChunkUtils.getLocalCoord;

public class ClientLevel extends Level{
    
    private final Minecraft session;
    private final ClientChunkManager chunkManager;
    private final ClientLevelConfiguration configuration;
    
    public ClientLevel(Minecraft session, String levelName){
        this.session = session;
        this.chunkManager = new ClientChunkManager(this);
        this.configuration = new ClientLevelConfiguration();
        
        configuration.load(levelName);
    }
    
    public Minecraft getSession(){
        return session;
    }
    
    
    @Override
    public short getBlock(int x, int y, int z){
        final ClientChunk targetChunk = getBlockChunk(x, z);
        if(targetChunk != null)
            return targetChunk.getBlock(getLocalCoord(x), y, getLocalCoord(z));
        
        return Blocks.VOID_AIR.getDefaultData();
    }
    
    @Override
    public boolean setBlock(int x, int y, int z, short block){
        final ClientChunk targetChunk = getBlockChunk(x, z);
        if(targetChunk != null)
            return targetChunk.setBlock(getLocalCoord(x), y, getLocalCoord(z), block);

        return false;
    }
    
    @Override
    public int getHeight(int x, int z){
        final ClientChunk targetChunk = getBlockChunk(x, z);
        if(targetChunk != null)
            return targetChunk.getHeightMap(HeightmapType.HIGHEST).getHeight(getLocalCoord(x), getLocalCoord(z));
        
        return 0;
    }
    
    
    @Override
    public byte getLight(int x, int y, int z){
        final ClientChunk targetChunk = getBlockChunk(x, z);
        if(targetChunk != null)
            return targetChunk.getLight(getLocalCoord(x), y, getLocalCoord(z));
        
        return 15;
    }
    
    @Override
    public void setLight(int x, int y, int z, int level){
        final ClientChunk targetChunk = getBlockChunk(x, z);
        if(targetChunk != null)
            targetChunk.setLight(getLocalCoord(x), y, getLocalCoord(z), level);
    }
    
    
    @Override
    public ClientLevelConfiguration getConfiguration(){
        return configuration;
    }
    
    @Override
    public ClientChunkManager getChunkManager(){
        return chunkManager;
    }
    
    @Override
    public ClientChunk getBlockChunk(int blockX, int blockZ){
        return chunkManager.getChunk(getChunkPos(blockX), getChunkPos(blockZ));
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/level/ClientLevel.kt
package pize.tests.voxelgame.client.level

import pize.tests.voxelgame.Minecraft
import pize.tests.voxelgame.client.block.Blocks
import pize.tests.voxelgame.client.chunk.ClientChunk
import pize.tests.voxelgame.main.chunk.ChunkUtils
import pize.tests.voxelgame.main.chunk.storage.HeightmapType
import pize.tests.voxelgame.main.level.Level

class ClientLevel(val session: Minecraft, levelName: String?) : Level() {
    override val chunkManager: ClientChunkManager
    override val configuration: ClientLevelConfiguration

    init {
        chunkManager = ClientChunkManager(this)
        configuration = ClientLevelConfiguration()
        configuration.load(levelName)
    }

    override fun getBlock(x: Int, y: Int, z: Int): Short {
        val targetChunk = getBlockChunk(x, z)
        return targetChunk?.getBlock(ChunkUtils.getLocalCoord(x), y, ChunkUtils.getLocalCoord(z))
            ?: Blocks.VOID_AIR.defaultData
    }

    override fun setBlock(x: Int, y: Int, z: Int, block: Short): Boolean {
        val targetChunk = getBlockChunk(x, z)
        return targetChunk?.setBlock(
            ChunkUtils.getLocalCoord(x),
            y,
            ChunkUtils.getLocalCoord(z),
            block
        )
            ?: false
    }

    override fun getHeight(x: Int, z: Int): Int {
        val targetChunk = getBlockChunk(x, z)
        return if (targetChunk != null) targetChunk.getHeightMap(HeightmapType.HIGHEST)!!
            .getHeight(ChunkUtils.getLocalCoord(x), ChunkUtils.getLocalCoord(z)) else 0
    }

    override fun getLight(x: Int, y: Int, z: Int): Byte {
        val targetChunk = getBlockChunk(x, z)
        return targetChunk?.getLight(ChunkUtils.getLocalCoord(x), y, ChunkUtils.getLocalCoord(z)) ?: 15
    }

    override fun setLight(x: Int, y: Int, z: Int, level: Int) {
        val targetChunk = getBlockChunk(x, z)
        targetChunk?.setLight(ChunkUtils.getLocalCoord(x), y, ChunkUtils.getLocalCoord(z), level)
    }

    override fun getBlockChunk(blockX: Int, blockZ: Int): ClientChunk? {
        return chunkManager.getChunk(ChunkUtils.getChunkPos(blockX), ChunkUtils.getChunkPos(blockZ))
    }
}
